### DISCUSSION TITLE
Set the constant frame rate when downloading the stream

### DISCUSSION THREAD

**Discussion by philyoun** [streamlink/streamlink#4983] [Q&A] [answered]
Hi, is there any way to set the constant frame rate when downloading the stream?

For example, i could handle this problem with
```
streamlink <url> best -o <downloaded_path>
ffmpeg -i <downloaded_path> -r 30 <another_path>
```
like this cause now I get the video with constant frame.

But is there another way to do this **at once** when download?

I checked the streamlink command-line usage to no avail.

Not like `streamlink <url> best -o <downloaded_path> | ffmpeg -i <downloaded_path> -r 30 <another_path>` cause this makes 2 jobs.

The reason I need this is, only with the video with constant frame, I can load this video on another program.
Any help will be appreciated. Thank you!

**Comment by bastimeyer** [comment, marked as answer]
> `streamlink <url> best -o <downloaded_path> | ffmpeg -i <downloaded_path> -r 30 <another_path>`

This is not how piping works. You're trying to execute two commands after another, so the pipe doesn't make any sense here, because Streamlink's stdout in this case is its log output, which gets piped into the ffmpeg process, and this is nonsense. `cmd1; cmd2` would execute `cmd1` and `cmd2` regardless the return code of `cmd1`, whereas `cmd1 && cmd2` would only execute `cmd2` if `cmd1` was successful, but that's not what you want either.

Take a look at Streamlink's available output parameters:

[`--output`/`-o`](https://streamlink.github.io/cli.html#cmdoption-output) writes the stream to the file system.
[`--stdout`/`-O`](https://streamlink.github.io/cli.html#cmdoption-stdout) on the other hand lets you pipe the stream into a process, like this:

So if you want to pipe the stream directly into ffmpeg, then do this

```
streamlink --stdout ... | ffmpeg -i pipe:0 ...
```

However, depending on the kind of stream and the stream container, this is not always possible (e.g. missing MPEG moov atom), so writing to the filesystem first may be required in some cases.

**Reply by philyoun** [reply]
OK, I misunderstood the pipe. Now I understood why that didn't even work.
But put aside the piping two commands, is there any way to download the stream with the constant frame?
I'm even thinking modify the `argparser.py` in streamlink.cli and `session.py` in streamlink...

**Comment by philyoun** [comment]
> This has been explained more than dozen times...
https://github.com/streamlink/streamlink/discussions/4541#discussion-4090978
